pkg/asset/installconfig/gcp/usertags.go

```diff
@@ -0,0 +1,247 @@
+}
+
+// validateAndPersistUserTags validates user provided tags are accessible and exists.
+// Converts and persists user-defined tags in NamespacedName
+// `{parentID}/{tag_key_short_name}/{tag_value_short_name}` form to key:value pairs,
+// with key of the form `tagKeys/{tag_key_id}` and value of the form
+// `tagValues/{tag_value_id}`. Returns error when fetching a tag fails or when
+// tag already exists on the project resource.
+func validateAndPersistUserTags(ctx context.Context, mgr TagManager, project string, userTags []gcp.UserTag) error {
+	if len(userTags) == 0 {
+		return nil
+	}
+
+	start := time.Now()
+	logrus.Debugf("user defined tags list: %v", userTags)
+
+	if len(userTags) > maxUserTagLimit {
+		return fmt.Errorf("more than %d user tags is not allowed, configured count: %d", maxUserTagLimit, len(userTags))
+	}
+
+	projectTags, err := mgr.GetProjectTags(ctx, project)
+	if err != nil {
+		return err
+	}
+
+	if dupTags := findDuplicateTags(userTags, projectTags); len(dupTags) != 0 {
+		return fmt.Errorf("found duplicate tags, %v tags already exist on %s project resource", dupTags, project)
+	}
+
+	processedTags.make(len(userTags))
+	nonexistentTags := make([]string, 0)
+	for _, tag := range userTags {
+		name := fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		tagValue, err := mgr.GetNamespacedTagValue(ctx, name)
+		if err != nil {
+			// check and return all non-existing tags at once
+			// for user to fix in one go.
+			var gErr *apierror.APIError
+			// google API returns StatusForbidden or StatusNotFound when the tag
+			// does not exist, since it could be because of permission issues
+			// or genuinely tag does not exist.
+			if errors.As(err, &gErr) && gErr.HTTPCode() == http.StatusForbidden {
+				logrus.Debugf("does not have permission to access %s tag or does not exist: %d", name, gErr.HTTPCode())
+				nonexistentTags = append(nonexistentTags, name)
+				continue
+			}
+			return fmt.Errorf("failed to fetch user-defined tag %s(%d): %w", name, gErr.HTTPCode(), err)
+		}
+		processedTags.addTag(tagValue.Parent, tagValue.Name)
+	}
+
+	if len(nonexistentTags) != 0 {
+		return fmt.Errorf("does not have permission to access %v tag(s) or does not exist", nonexistentTags)
+	}
+
+	processedTags.markProcessed()
+	logrus.Debugf("user defined tags processed list: %v, took %s to finish", processedTags.tags, time.Since(start))
+	return nil
+}
+
+// findDuplicateTags returns list of duplicate userTags which are already present
+// in the parentTags list.
+func findDuplicateTags(userTags []gcp.UserTag, parentTags sets.Set[string]) []string {
+	dupTags := make([]string, 0)
+	for _, tag := range userTags {
+		tagNamespacedName := fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		if parentTags.Has(tagNamespacedName) {
+			dupTags = append(dupTags, tagNamespacedName)
+		}
+	}
+	return dupTags
+}
+
+// getCloudResourceServiceForTags returns the client required for querying resource manager resources.
```

Comment:
[Code](https://github.com/openshift/installer/pull/7279/files#diff-c4351fb8fcc08cb33821af969413576f521211960eeeb7c0c299ffa2e66cefc1R192-R247) will be merged with client code in `pkg/asset/installconfig/gcp/client.go` after updating `google.golang.org/api/cloudresourcemanager` from `v1` to `v3` in the follow up PR.